src/app/blog/irpc/page.mdx

```diff
@@ -0,0 +1,347 @@
+
+```rust
+enum Client {
+  Local(mpsc::Sender<FullRequest>),
+  Remote(quinn::Connection),
+}
+
+impl Client {
+  async fn get(&self, key: String) -> Result<Option<String>> {
+    let request = GetRequest { key };
+    match self {
+      Self::Local(chan) => {
+        let (tx, rx) = oneshot::channel();
+        let request = FullRequest { request, response: tx };
+        chan.send(request).await?;
+        Ok(rx.await??)
+      }
+      Self::Remote(conn) => {
+        let (send, recv) = connection.open_bi().await?;
+        send.write_all(postcard::to_stdvec(request)?).await?;
+        let res = recv.read_to_end(1024).await?;
+        let res = postcard::from_bytes(&res)?;
+        Ok(res)
+      }
+    }
+  }
+}
+```
+
+This is all pretty straightforward code, but very tedious to write, especially for a large and complex protocol.
+
+There is some work that we can't avoid. We have to define the different request types. We have to specify for each request type the kind of response we expect (no response, a single response, or a stream of responses). We also have to specify if there are updates and make sure that all these types (requests, updates and responses) are serializable, which can sometimes be a pain when it comes to error types.
+
+But what about all this boilerplate?
+- Defining the two different enums for a serializable request and a full request including channels
+- Implementing a client with async fns for each request type
+- Implementing a server that reads messages and dispatches on them
+- serializing and deserializing using postcard with length prefixes
+
+**The `irpc` crate is meant solely to reduce the tedious boilerplate involved in writing the above manually.**
+
+It does *not* abstract over the connection type - it only supports [iroh-quinn] send and receive streams out of the box, so the only two possible connection types are `iroh` p2p QUIC connections and normal QUIC connections. It also does not abstract over the local channel type - a local channel is always a `tokio::sync::mpsc` channel. Serialization is always using postcard and length prefixes are always postcard varints.
+
+So let's see what our kv service looks using `irpc`:
+
+The service definition contains just what is absolutely needed. For each request type we have to define what the response item type is (in this case `String` or `()`), and what the response channel type is (none, oneshot or mpsc).
+
+The `rpc_requests` macro will store this information and also create the `RequestWithChannels` enum that adds the appropriate channels for each request type. It will also generate a number of `From`-conversions to make working with the requests more pleasant.
+
+```rust
+struct KvService {}
+impl Service for KvStoreService {}
+
+#[rpc_requests(KvService, message = RequestWithChannels)]
+#[derive(Serialize, Deserialize)]
+enum Request {
+  #[rpc(tx=oneshot::Sender<String>)]
+  Get(GetRequest),
+  #[rpc(tx=oneshot::Sender<()>)]
+  Put(PutRequest),
+}
```

Comment:
I think it'd be good to add an import line for these `oneshot` modules, or perhaps even make them fully qualified, since it's very confusing that these are actually from `irpc`, not from `tokio` or `std` (what one would otherwise expect).